src/aapns/connection.py

```diff
@@ -0,0 +1,454 @@
+                            # Stream flow control is responsibility of the channel.
+                            # Connection flow control is handled here.
+                            self.protocol.increment_flow_control_window(
+                                event.flow_controlled_length
+                            )
+                        if channel:
+                            channel.events.append(event)
+                            channel.wakeup.set()
+
+                # Somewhat inefficient: wake up background writer just in case
+                # it could be that we've received something that h2 needs to acknowledge
+                self.should_write.set()
+
+                # FIXME notify pool users about possible change to `.blocked`
+                # * h2.events.WindowUpdated
+                # * max_concurrent_streams change
+                # * [maybe] starting a stream
+                # * a stream getting closed (but not half-closed)
+                # * closing / closed change
+        except ConnectionError as e:
+            if not self.outcome:
+                self.outcome = str(e)
+        except Exception:
+            logger.exception("background read task died")
+        finally:
+            self.closing = self.closed = True
+            for channel in self.channels.values():
+                channel.wakeup.set()
+
+    async def background_write(self):
+        try:
+            while not self.closed:
+                data = None
+
+                while not data:
+                    if self.closed:
+                        return
+
+                    if data := self.protocol.data_to_send():
+                        self.write_stream.write(data)
+                        last_stream_id = self.last_stream_id_got
+                        await self.write_stream.drain()
+                        self.last_stream_id_sent = last_stream_id
+                    else:
+                        await self.should_write.wait()
+                        self.should_write.clear()
+
+        except (SSLError, ConnectionError) as e:
+            if not self.outcome:
+                self.outcome = str(e)
+        except Exception:
+            logger.exception("background write task died")
+        finally:
+            self.closing = self.closed = True
+
+
+@dataclass
+class Channel:
+    wakeup: asyncio.Event = field(default_factory=asyncio.Event)
+    events: List[h2.events.Event] = field(default_factory=list)
+    header: Optional[dict] = None
+    body: bytes = b""
+
+
+@dataclass
+class Request:
+    header: tuple
+    body: bytes
+    deadline: float
+
+    def header_with(self, host: str, port: int) -> tuple:
+        """Request header including :authority pseudo header field for target server"""
+        return ((":authority", f"{host}:{port}"),) + self.header
+
+    @classmethod
+    def new(
+        cls,
+        path: str,
+        header: Optional[dict],
```

Comment:
```suggestion
        headers: Optional[Dict[str,str]],
```

Reply:
if it's after `data` it can also get a `=None`

Reply:
APN has mandatory header fields, I'm making the header mandatory too now.